service/history/queues/executable.go

```diff
@@ -506,6 +513,23 @@ func (e *executableImpl) HandleErr(err error) (retErr error) {
 		}
 	}()
 
+	dlqErrSubStrings := e.dlqErrorSubStrings()
```

Comment:
can we only run this parsing & matching logic later when we can't handle the error? Like at the end of `isUnexpectedNonRetryableError()`. The logic also seems to belong to that method.

Reply:
Is there any reason we'd want to use this to override all existing error handling behavior? If so it should happen up front

Reply:
Putting this here will give us the option to even catch retryable errors if we need that in the future.  Do we not want that flexibility?

Reply:
> catch retryable errors, SafeToDropErrors

hmm I thought the intention of the change is to be able to dlq unknown errors sooner without waiting for the attempts to exceed threshold. Override existing behavior, hmmm, maybe, though I don't see a clear use case now. Tracking workflows impacted by a bug where tasks are wrongly dropped, maybe. Not quite following the Testing use case. If we want to trigger a task to go to dlq, then the test can just use a non-retryable error. 
 
I guess I am mainly worried about the additional cost for string parsing and regex matching logic especially during throttled case where tons of task getting resource exhausted errors from downstream and evaluate the error handling logic here.



> If the error matches with isUnexpectedNonRetryableError() it will anyway go to dlq.

I mean this string matching logic is for deciding if error is unexpected and non-retryable and should go to dlq, so is doing the same thing as the rest of the logic in `isUnexpectedNonRetryableError` so I think logically belong there.

Reply:
**RE: string parsing/regex compilation logic**

What'd be nice is if we could _only_ re-evaluate `e.dlqErrorRegex()` when it changes. There's no reason to call that function and compile the regular expression on every change.

If we're concerned we could cache the last-compiled regex and use it if the string patterns match. It's not something that will change often

Reply:
I thought of something like, we expected an error to be retryable and we then realized this is not going to succeed. We can use this config to send that task to dlq.

We will only make this comparison if dlqErrorRegex() is provided. In normal cases, we will not have this config set. So no performance overhead. We can set it to catch some failing task and then reset it soon after that.

But if we want to set this config and run the cluster for a long time, then there can be a performance issue. It is better to move it down.

Reply:
> We will only make this comparison if dlqErrorRegex() is provided. In normal cases, we will not have this config set. So no performance overhead.

Ah right, makes sense. I think under normal cases we should not have this flag set. 

> only re-evaluate e.dlqErrorRegex() when it changes

Our dynamic config collection doesn't support value change subscription right now. It's something I want to do for a long time but never really got the chance.

Reply:
I checked with @yiminc as well. He also shared the opinion that we should be able to catch retryable error if needed and we will not set this config unless we need, so we don't have to worry that much about performance.

Reply:
@yycptt we can cache the previous value ourselves and do a simple strcmp to see if they differ :). That'll be faster than recompiling the regex on every iteration, and is less code than, say, a compiled pattern cache

Reply:
@tdeebswihart This function is inside the Executable object. This object is created for every task. So we cannot cache it inside this object.

Also, if we want to cache this somewhere else, we must use a lock around that code. All tasks will now have to wait on this lock. So it is a possible contention point. Is it worth doing that? given that we intend to set this config for short periods.

Reply:
That’s unfortunate but it probably won’t matter. If for some reason this is identified as a bottleneck we could cache it but until then let’s punt